### PR TITLE
Add tests with switch expressions

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AllBranchesShouldNotHaveSameImplementation.cs [hunk 1]
```diff
@@ -144,5 +144,17 @@ private void DoSomething()
         private void DoSomethingElse()
         {
         }
+
+        public int Test(string type)
+        {
+            return type switch // Compliant - FN: all branches call the same function
+            {
+                "a" => GetNumber(),
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/AllBranchesShouldNotHaveSameImplementation.cs]: would it be useful to add
- a test with nested identical switch expressions?
- compliant tests for switch expressions

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_CustomValue.cs [hunk 2]
```diff
@@ -22,5 +22,21 @@ public Program(int myVariable)
                     break;
             }
         }
+
+        public int Test(string type)
+        {
+            return type switch // Compliant - FN
+            {
+                _ => 1
+                       + 2
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchSectionShouldNotHaveTooManyStatements_CustomValue.cs]: this is not really multiple statements, but a big expression inside - I think we don't have a rule for that ("this statement has too many expressions in it"), and it's out of scope anyway

so in this case, the rule does not really apply to switch expressions (given they cannot have statements inside)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchShouldNotBeNested.cs [hunk 3]
```diff
@@ -42,6 +42,19 @@ public Program(char choice, int value)
                     Console.WriteLine("Other");
                     break;
             }
+
+            string first = "foo", second = "bar";
+
+            var result = first switch
+            {
+                "a" => second switch // Compliant - FN: should check also switch expressions?
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchShouldNotBeNested.cs]: I'm not sure it should apply to switch expressions... at least MS gives example w/ nested switch expressions on their docs site, like a pattern that can be used. 

We can discuss whether it's a nice pattern or not, but nested switch expressions are less bad then nested switch statements (where you can have break, early return, multiple statements per case). Switch expressions have a single return.

@nicolas-harraudeau-sonarsource - WDYT?

Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/SwitchShouldNotBeNested.cs]: Indeed, nested expressions are hard to read but not as bad as the statements. 

I've removed the "FN" comment but I've let the test to document the current behavior. We can decide later if we want to extend the rule to cover also this.

**Pull request merged**